### PR TITLE
Add `capture` support for unconventional result types

### PR DESCRIPTION
**Pull request by kennykerr**
Fixes: #1299

### CODE REVIEW

**Review by jonwis — DISMISSED** [review]

File: strings/base_com_ptr.h [hunk 1]
```diff
@@ -7,16 +7,27 @@ WINRT_EXPORT namespace winrt
 
 namespace winrt::impl
 {
+    struct capture_decay
+    {
+        void** result;
+
+        template <typename T>
+        operator T** ()
```
Comment (sylveon) [inline strings/base_com_ptr.h:15]: Maybe check if `T` derives from IUnknown?

File: test/old_tests/UnitTests/capture.cpp [hunk 2]
```diff
@@ -41,6 +48,12 @@ HRESULT __stdcall CreateCapture(int value, GUID const& iid, void** object) noexc
     return capture->QueryInterface(iid, object);
 }
 
+HRESULT __stdcall CreateCapture2(int value, GUID const& iid, ::IUnknown** object) noexcept
```
Comment (sylveon) [inline test/old_tests/UnitTests/capture.cpp:51]: IInspectable was mentionned in the original issue. Sounds fair to test it too.

File: strings/base_com_ptr.h (hunk 1 above)
Comment (kennykerr) [reply strings/base_com_ptr.h:15]: In most cases `T` will be `void` - the whole point here is to throw away type safety and just take whatever is provided, as long as it's a `T**`.

**Comment by oldnewthing** [comment]
I was originally worried that this might be too permissive. There are some methods that take a REFIID followed by something that *isn't* the matched output pointer. However, they are exceedingly rare, and the only one that does it at the end of the parameter list that I could find (mind you, I gave up halfway through the D's) is IDebugProperty::EnumMembers.

Unit test needs to include the forwarder pattern:

```cpp
com_ptr<ICapture> f = capture<ICapture>([&](auto riid, auto ppv) { return CreateCapture2(10, riid, ppv); });
```

This pattern is used to accommodate methods that don't put the riid/ppv as the last two parameters.

```cpp
HRESULT WeirdMethod(REFIID riid, char const* interloper, IUnknown** result, int straggler);

e.capture<IUnknown>(WeirdMethod, [](auto interloper, auto straggler, riid, auto result) { return WeirdMethod(riid, interloper, result, straggler); }, "interloper", 10);
```

**Pull request merged**